pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,433 @@
+// the traffic through tunnels between Gateway and other Nodes, as otherwise the
+// traffic will not go through tunnels in those modes.
+type MCPodRouteController struct {
+	k8sClient       kubernetes.Interface
+	mcClient        mcclientset.Interface
+	ovsBridgeClient ovsconfig.OVSBridgeClient
+	ofClient        openflow.Client
+	interfaceStore  interfacestore.InterfaceStore
+	nodeConfig      *config.NodeConfig
+	podQueue        workqueue.RateLimitingInterface
+	gwQueue         workqueue.RateLimitingInterface
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	gwInformer      cache.SharedIndexInformer
+	gwLister        mclisters.GatewayLister
+	namespace       string
+	// podCache stores Pods' latest IP and Node IP.
+	podCache      map[string]*podNodeIPs
+	podCacheMutex sync.RWMutex
+	// ipToPods stores NamespacedNames of all Pods with the same IP.
+	ipToPods      map[string]sets.String
+	ipToPodsMutex sync.RWMutex
+	// podWorkersStarted is a boolean which tracks if the Pod flow controller has been started.
+	podWorkersStarted      bool
+	podWorkersStartedMutex sync.RWMutex
+	podWorkerStopCh        chan struct{}
+}
+
+func NewMCPodRouteController(
+	k8sClient kubernetes.Interface,
+	mcClient mcclientset.Interface,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		k8sClient:       k8sClient,
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		podQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForPod"),
+		gwQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForGateway"),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		podCache:        make(map[string]*podNodeIPs),
+		ipToPods:        make(map[string]sets.String),
+		podWorkerStopCh: make(chan struct{}),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) createPodInformer() {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", c.nodeConfig.Name).String()
```

Comment:
I also have a question on this one - what will happen if the Pod is moved (rescheduled) from another Node to this Node. Will we always get a Pod deletion for that if we have this filter? @tnqn